### PR TITLE
Handle aliases/typedefs in shaper functions

### LINKED ISSUE #2192: Make shapers work with typedefs
Address @mccanne's comments in https://github.com/brimsec/zq/pull/2173 (also if relevant for the prior shaper work).

### LINKED ISSUE #2219: "panic: interface conversion" while trying to shape
Repro is with `zq` commit `1ac7787f`.

I was trying to verify #2198 and was hacking around before fully understanding the example, and triggered a panic.

```
$ cat example.zs 
type port=uint16;
type conn={orig_h:ip,orig_p:port,resp_h:ip,resp_p:port};
put . = shape(., {_path:string,ts:time,uid:bstring,id:conn})

$ echo '{"foo": "bar"}' | zq -z -I example.zs -
panic: interface conversion: zng.Type is *zng.TypeAlias, not *zng.TypeRecord
goroutine 68 [running]:
runtime/debug.Stack(0x46, 0x40, 0x6b)
	/usr/local/opt/go/libexec/src/runtime/debug/stack.go:24 +0x9f
github.com/brimsec/zq/zqe.RecoverError(0x182e8c0, 0xc0003bc690, 0xc0003bc690, 0xc000616200)
	/Users/phil/work/zq/zqe/error.go:163 +0x26
github.com/brimsec/zq/driver.(*mux).safeGet.func1(0xc000625f70)
	/Users/phil/work/zq/driver/mux.go:49 +0x48
panic(0x182e8c0, 0xc0003bc690)
	/usr/local/opt/go/libexec/src/runtime/panic.go:969 +0x1b9
github.com/brimsec/zq/expr.(*Shaper).orderRecordType(0xc000132360, 0xc0003bc300, 0xc0003bc300, 0xc0003bc300, 0x0, 0x0)
	/Users/phil/work/zq/expr/shaper.go:404 +0xa10
github.com/brimsec/zq/expr.(*Shaper).createShapeSpec(0xc000132360, 0xc0003bc5a0, 0xc0003bc300, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/phil/work/zq/expr/shaper.go:323 +0x1ee
github.com/brimsec/zq/expr.(*Shaper).Eval(0xc000132360, 0xc0000325f0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc000032501, 0xc000128c60)
	/Users/phil/work/zq/expr/shaper.go:285 +0x337
github.com/brimsec/zq/proc/put.(*Proc).eval(0xc0000a62d0, 0xc0000325f0, 0x0, 0xc000128c60, 0xc000625d98, 0x15b87dc, 0xc000128080)
	/Users/phil/work/zq/proc/put/put.go:84 +0x127
github.com/brimsec/zq/proc/put.(*Proc).put(0xc0000a62d0, 0xc0000325f0, 0xc0000325f0, 0xc000010040, 0x0)
	/Users/phil/work/zq/proc/put/put.go:347 +0x50
github.com/brimsec/zq/proc/put.(*Proc).Pull(0xc0000a62d0, 0x0, 0x0, 0x0, 0x0)
	/Users/phil/work/zq/proc/put/put.go:371 +0x15f
github.com/brimsec/zq/driver.(*mux).safeGet(0xc0001282a0, 0x0, 0x0, 0x0, 0x0)
	/Users/phil/work/zq/driver/mux.go:51 +0x6c
github.com/brimsec/zq/driver.(*mux).run(0xc0001282a0)
	/Users/phil/work/zq/driver/mux.go:64 +0x32
created by github.com/brimsec/zq/driver.(*muxOutput).Pull.func1
	/Users/phil/work/zq/driver/mux.go:98 +0x5d
```

### PR DESCRIPTION
**Pull request by henridf**
closed #2192 
fixes #2219

### CODE REVIEW

File: expr/shaper.go [hunk 1]
```diff
@@ -398,18 +407,22 @@ func (s *Shaper) orderRecordType(input, spec *zng.TypeRecord) (*zng.TypeRecord,
...
 					return nil, err
 				} else {
 					cols = append(cols, zng.Column{specCol.Name, nested})
 				}
 				continue
 			}
 			if isCollectionType(inCol.Type) && isCollectionType(specCol.Type) && zng.IsRecordType(innerType(inCol.Type)) && zng.IsRecordType(innerType(specCol.Type)) {
-				if inner, err := s.orderRecordType(innerType(inCol.Type).(*zng.TypeRecord), innerType(specCol.Type).(*zng.TypeRecord)); err != nil {
+				inInner := zng.AliasedType(innerType(inCol.Type))
+				specInner := zng.AliasedType(innerType(specCol.Type))
```
Comment (nwt) [inline expr/shaper.go:424]:
```suggestion
				inInner := zng.AliasedType(zng.InnerType(zng.AliasedType(inCol.Type)))
				specInner := zng.AliasedType(zng.InnerType(zng.AliasedType(specCol.Type)))
```

File: expr/shaper.go [hunk 2]
```diff
@@ -460,21 +475,25 @@ func (s *Shaper) fillRecordType(input, spec *zng.TypeRecord) (*zng.TypeRecord, e
...
-					var t zng.Type
-					if _, ok := inCol.Type.(*zng.TypeArray); ok {
-						t, err = s.zctx.LookupTypeArray(inner), nil
-					} else {
-						t, err = s.zctx.LookupTypeSet(inner), nil
-					}
-					if err != nil {
+			if isCollectionType(inType) && isCollectionType(specType) {
+				inInner := zng.AliasedType(innerType(inCol.Type))
+				specInner := zng.AliasedType(innerType(specCol.Type))
```
Comment (nwt) [inline expr/shaper.go:480]:
```suggestion
				inInner := zng.AliasedType(zng.InnerType(zng.AliasedType(inCol.Type)))
				specInner := zng.AliasedType(zng.InnerType(zng.AliasedType(specCol.Type)))
```

**Review by nwt — APPROVED** [review]

File: expr/shaper.go [hunk 3]
```diff
@@ -93,17 +93,19 @@ func createStepSet(in, out zng.Type) (step, error) {
...
+	switch t := t.(type) {
+	case *zng.TypeAlias:
+		return isCollectionType(t.Type)
+	case *zng.TypeArray, *zng.TypeSet:
 		return true
 	}
 	return false
 }
 
 func innerType(t zng.Type) zng.Type {
```
[the remark this answers is not in the document]
Comment (henridf) [reply expr/shaper.go:105]: I agree but this also makes me wonder if zng.InnerType shouldn't be changed to unwrap aliases. I filed #2270 for this and will leave this in the meantime..

File: expr/shaper.go [hunk 4]
```diff
@@ -340,37 +342,44 @@ func (s *Shaper) cropRecordType(input, spec *zng.TypeRecord) (*zng.TypeRecord, e
...
-				return nil, err
-			}
-			var t zng.Type
-			if _, ok := inCol.Type.(*zng.TypeArray); ok {
-				t, err = s.zctx.LookupTypeArray(inner), nil
+		case isCollectionType(inType) && isCollectionType(specType):
+			if zng.IsRecordType(innerType(inType)) && zng.IsRecordType(innerType(specType)) {
+				// 4. array/set of records
+				inInner := zng.AliasedType(innerType(inType))
+				specInner := zng.AliasedType(innerType(specType))
```
[the remark this answers is not in the document]
Comment (henridf) [reply expr/shaper.go:363]: 👍

File: expr/shaper.go [hunk 5]
```diff
@@ -398,18 +407,22 @@ func (s *Shaper) orderRecordType(input, spec *zng.TypeRecord) (*zng.TypeRecord,
 	// with use, so starting with simpler algorithm for now.
 	//
 	for _, specCol := range spec.Columns {
+		specType := zng.AliasedType(specCol.Type)
```
[the remark this answers is not in the document]
Comment (henridf) [reply expr/shaper.go:410]: 👍

File: expr/shaper.go (hunk 1 above)
Comment (henridf) [reply expr/shaper.go:424]: https://github.com/brimsec/zq/pull/2257#discussion_r585338529

**Pull request merged**